### DISCUSSION TITLE
glb/gltf Transformations, Rotations and Scaling

### DISCUSSION THREAD

**Discussion by hanan-commits-work** [iTwin/itwinjs-core#3279] [Q&A] [answered]
Greetings.
I added the .glb model in the app using local file storage. Now, I wanted to add the following capabilities...
1. Select Model onClick
2. Change Positioning of the Model 
3. Rotate the Model along the axis
4. Scale the Model based on x,y,z coordinates of the model

If you have any sample apps or example code to help me achieve these capabilities it would be extremely helpful.
Thanks.

**Comment by pmconne** [comment, marked as answer]
Is your code based upon [this sample](https://www.itwinjs.org/sandboxes/NathanElrod/glTF%20Decorator)?

1: This should just work - the sample makes a pickable decoration. However I believe the sandbox currently has a bug that @StefanRetief and/or @Ellord207 is investigating wherein you first have to use the rotate tool before selecting the model works properly.

2, 3, 4: i.e., you want to apply a [Transform](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/transform/) to your decoration graphics. You can do so by placing the graphic inside a [GraphicBranch](https://www.itwinjs.org/reference/core-frontend/rendering/graphicbranch/).

If further assistance is required please supply your code.

**Reply by hanan-commits-work** [reply]
Thank you for the answer.
And yes I am making the code based on the excellent sample by @Ellord207 .. I will look  into the code
and here is the sample code in which i am working on.
Let me walk you through the code running process.

when you start the app, you will see a little green button on the left side menu. When you click on it, the placement tool will be activated and you just click anywhere on the screen. And then it will open a window and you can use it to select any glb file on your system. I also provided the model.glb file in the public folder of the code.
[zip.zip](https://github.com/iTwin/itwinjs-core/files/8153803/zip.zip)

**Reply by pmconne** [reply]
Are the answers provided sufficient?

**Reply by hanan-commits-work** [reply]
I am looking at the documentations you provided. Can you please give me an overview of how the Transform and Graphic Branch actually works?
Specifically, how does it adds my model and then creates its vector and then scales, rotates, etc.

**Reply by pmconne** [reply]
I [forked the sandbox](https://www.itwinjs.org/sandboxes/pmconne/Transform%20glTF%20decoration) to add rotation, translation, and scale.

**Reply by hanan-commits-work** [reply]
Thank you for this @pmconne . 
I am having trouble in handing ButtonEvent in the decorator. I am using the following approach, ( I understand that it maybe wrong so I would appreciate if you could correct my understanding).
TestDecorationHit returns true when the mouse is hovered on the model. Then I am trying to call onDecorationButtonEvent. But there is no Hit or Event args in the TestDecorationHit. So I cant figure out the way to handle my button events (onDown, onDoubleClick, etc)
Kindly let me know where I am wrong or if there is a workaround for this.

**Reply by hanan-commits-work** [reply]
I noticed that onDecorationButtonEvent is called whenever there is a click, but somehow, it is not being called in my code.

**Reply by hanan-commits-work** [reply]
I noticed that even in the sandbox (https://www.itwinjs.org/sandboxes/NathanElrod/glTF%20Decorator), onDecorationButtonEvent isnt being called. and then I followed this [sample code ](https://www.itwinjs.org/learning/frontend/viewdecorations/#gltf-decorations) but somehow clicks arent being recorded.

**Reply by hanan-commits-work** [reply]
any response in this?

**Reply by pmconne** [reply]
Please add an invalid modelId to your glTF graphics like so (the only change from the sample is to add `modelId: Id64.invalid`).
This will make the graphic selectable and invoke your onDecorationButtonEvent if testDecorationHit returns true. (You should not be trying to invoke onDecorationButtonEvent yourself).
If problems persist, post your code (or better, link to a sandbox illustrating the problem).
See [documentation](https://www.itwinjs.org/reference/core-frontend/tiles/readgltfgraphicsargs/pickableoptions/).
```
    let graphic = await readGltfGraphics({
      gltf: new Uint8Array(buffer),
      iModel,
      pickableOptions: {
        id,
        modelId: Id64.invalid,
      },
    });
```
